mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
-            .context(ReadError {
-                name: String::from("header"),
-            })?;
-        let header = NexusMetaData::read_header(&buf)?;
-
-        //
-        // Index
-        let index = if header.used_entries > 0 {
-            let blocks = Aligned::get_blocks(
-                (header.used_entries * header.entry_size) as u64,
-                block_size,
-            );
-            let mut buf =
-                hndl.dma_malloc(blocks * block_size).context(ReadAlloc {
-                    name: String::from("index"),
-                })?;
-            hndl.read_at(
-                (header.self_lba + header.index_start) * block_size,
-                &mut buf,
-            )
-            .await
-            .context(ReadError {
-                name: String::from("index"),
-            })?;
-            NexusMetaData::read_index(&buf, &header)?
-        } else {
-            NexusMetaData::empty_index(&header)?
-        };
+pub struct NexusMetaData;
 
-        Ok(NexusMetaData {
-            header,
-            index,
-        })
+impl NexusMetaData {
+    fn timestamp(time: &SystemTime) -> u128 {
+        time.duration_since(UNIX_EPOCH).unwrap().as_micros()
     }
 
-    /// Read the selected config object from disk.
-    async fn probe_config_object(
-        &self,
-        metadata: &NexusMetaData,
-        selected: u32,
-    ) -> Result<NexusConfig, MetaDataError> {
-        if selected >= metadata.header.used_entries {
-            return Err(MetaDataError::IndexOutOfRange {
-                selected,
-                used: metadata.header.used_entries,
-            });
-        }
+    fn read_index(buf: &DmaBuf) -> Result<MetaDataIndex, MetaDataError> {
+        MetaDataIndex::from_slice(buf.as_slice())
+    }
 
-        let entry = &metadata.index[selected as usize];
+    fn read_object(buf: &DmaBuf) -> Result<MetaDataObject, MetaDataError> {
+        MetaDataObject::from_slice(buf.as_slice())
+    }
 
-        let (bdev, hndl) = self.get_dev().context(NexusChildError {})?;
-        let block_size = bdev.block_len() as u64;
+    fn write_index(
+        buf: &mut DmaBuf,
+        index: &mut MetaDataIndex,
+        now: &SystemTime,
+    ) -> Result<(), MetaDataError> {
+        buf.fill(0);
+        let mut writer = Cursor::new(buf.as_mut_slice());
 
-        let blocks = entry.data_end - entry.data_start + 1;
-        let mut buf =
-            hndl.dma_malloc(blocks * block_size).context(ReadAlloc {
-                name: String::from("object"),
-            })?;
-        hndl.read_at(
-            (metadata.header.self_lba + entry.data_start) * block_size,
-            &mut buf,
-        )
-        .await
-        .context(ReadError {
-            name: String::from("object"),
-        })?;
+        index.generation += 1;
+        index.timestamp = NexusMetaData::timestamp(now);
+        index.checksum();
+        serialize_into(&mut writer, index).context(SerializeError {})
+    }
+
+    fn write_object(
```

Comment:
Im not sure if we have ever discussed this I do recall is talking about this subject but to be sure, how do we handle atomicity? 

What if the buffer we need to write exceeds the blk_size?